lib/includes.js

```diff
@@ -0,0 +1,189 @@
+            var branch = _.map(typeMap, function (value, key) {
+                path.push(key);
+                var subBranch = buildInclusionBranch([key], schema);
+                if (!_.isEmpty(value)) {
+                    if (isRemote(subBranch, key)) {
+                        subBranch.refs[key].def.remoteIncludes = toIncludes(value);
+                    } else {
+                        var schemaName = schema.relationships[key].type;
+                        var subMap = mapToInclusionTree(value, schemas[schemaName], path);
+                        _.merge(subBranch.refs[key], subMap);
+                    }
+                }
+                path.pop();
+                return subBranch;
+            });
+            return _.reduce(branch, function (acc, item) {
+                return _.merge(acc, item);
+            }, {});
+        }
+
+        var inclusionMap = _.map(inclusions, function (inclusionString) {
+            var result = {};
+            var node = result;
+            inclusionString.split('.').forEach(function (token) {
+                node = node[token] = {};
+            });
+            return result;
+        }).reduce(function (acc, node) {
+            return _.merge(acc, node);
+        }, {});
+
+        var inclusionTree = mapToInclusionTree(inclusionMap, primarySchema);
+        // builds a tree representation out of a series of inclusion tokens
+        function buildInclusionBranch(inclusionTokens, schema) {
+
+            var inclusionToken = _.first(inclusionTokens);
+            var type = _.isArray(schema.relationships[inclusionToken]) ? schema.relationships[inclusionToken][0] : schema.relationships[inclusionToken];
+            var normalisedType = _.isPlainObject(type) ? type.type : type;
+
+            var linkDescriptor = {
+                def: {type: normalisedType}
+            };
+
+            var baseUri = _.isPlainObject(type) ? type.baseUri : null;
+            if (baseUri) {
+                var remoteIncludes = _.drop(inclusionTokens, _.indexOf(inclusionTokens, inclusionToken) + 1).join('.');
+                var remoteDescriptor = _.merge(linkDescriptor, {
+                        def: {baseUri: baseUri, remoteIncludes: remoteIncludes}
+                    }
+                );
+                return setRefs(remoteDescriptor);
+            } else {
+                var tokensRemaining = _.drop(inclusionTokens);
+                if (tokensRemaining.length == 0) {
+                    return setRefs(linkDescriptor);
+                } else {
+                    return _.merge(linkDescriptor, buildInclusionBranch(tokensRemaining, schemas[normalisedType]));
+                }
+            }
+
+            function setRefs(descriptor) {
+                return _.set({}, 'refs.' + inclusionToken, descriptor);
+            }
+
+        }
+
+        var resources = _.isArray(body.data) ? body.data : [body.data];
+        return fetchLinked({}, resources, inclusionTree)
```

Comment:
I sort of feel like we're using too many names to describe the same thing? ```relationship```, ```include``` and ```links```